Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -224,6 +224,22 @@ final class ConnectionPool {
         fileprivate var closePromise: EventLoopPromise<Void>
 
         var closeFuture: EventLoopFuture<Void>
+
+        func removeIdleConnectionHandlersForLease() -> EventLoopFuture<Connection> {
+            return self.channel.eventLoop.flatSubmit {
+                self.removeHandler(IdleStateHandler.self).flatMap {
```

Comment:
this is _almost_ correct but there's still a tiny little race left unfortunately (which I didn't think of before): It's possible that the `IdlePoolConnectionHandler` has already issued a `close` but the `close` might not actually have happened yet. Why is that? It's because the `SSLHandler` actually has to actually do a "send a `CloseNotify`/wait for `CloseNotify`" dance to close a connection. So with an `SSLHandler` in the pipeline, a `close` unfortunately won't complete synchronously. So it's possible that a `Channel` is both unusable and yet `channel.isActive == true`.

So I propose two things:
1. let's write a test that highlights this (I'll help with that one in a second)
2. fix the race. The only way I can think of to fix this is to add _one more_ handler at the front which is a `ChannelDuplexHandler`. It holds the state whether the channel is usable. It does this by: listening for `channelInactive` (when received: unusable) and also offering an outbound user event `TearDownChannelEvent` or so. When it receives `TearDownChannelEvent`, this handler will a) mark the channel as unusable b) call `context.channel.close()`.

Then, the only thing that's left is to query the `TearDownChannelEvent` handler's state before deciding whether this channel is usable or not.

Does this make some sense? If not, please ping on Discord or so.

Reply:
@adtrevor https://github.com/adtrevor/async-http-client/pull/4

Reply:
@weissi Thank you!

Reply:
@weissi Currently the pool waits for a connection to actually close before creating the new connection, so  this means in your test `doActualCloseNowPromise.succeed(())` cannot come after request execution, I modified for now, but then it makes it non-deterministic 😕

I think we can either consider that the code is ok as is and in such case use only non-deterministic tests, or if we consider the current behaviour is wrong re-factor the code to avoid waiting on the connection to close.

Reply:
The latest commit should fix the race 🙂

Reply:
@adtrevor looks great! One small issue but then we can merge